packages/bsky/src/cache/read-through.ts

```diff
@@ -0,0 +1,131 @@
+import { CacheItem, RedisCache } from './redis'
+import { cacheLogger as log } from '../logger'
+
+export type CacheOptions<T> = {
+  staleTTL: number
+  maxTTL: number
+  namespace?: string
+  fetchMethod: (key: string) => Promise<T | null>
+  fetchManyMethod?: (keys: string[]) => Promise<Record<string, T | null>>
+}
+
+export class ReadThroughCache<T> {
+  constructor(public redisCache: RedisCache, public opts: CacheOptions<T>) {}
+
+  private async _fetchMany(keys: string[]): Promise<Record<string, T | null>> {
+    if (this.opts.fetchManyMethod) {
+      return this.opts.fetchManyMethod(keys)
+    }
+    const got = await Promise.all(keys.map((k) => this.opts.fetchMethod(k)))
+    const result: Record<string, T | null> = {}
+    for (let i = 0; i < keys.length; i++) {
+      result[keys[i]] = got[i] ?? null
+    }
+    return result
+  }
+
+  private async fetchAndCache(key: string): Promise<T | null> {
+    const fetched = await this.opts.fetchMethod(key)
+    this.set(key, fetched).catch((err) =>
+      log.error({ err, key }, 'failed to set cache value'),
+    )
+    return fetched
+  }
```

Comment:
yea i thought about trying to get some coalescing in, but it seemed prone to errors & this is gonna be split across ~40 instances so i'm not sure how much it'll help

since this is technically correct now, think i'm gonna leave as is unless we notice issues with it